src/signing/collector/signatures_collector.rs

```diff
@@ -0,0 +1,554 @@
+        )
+    }
+}
+
+// === PUBLIC ===
+impl SignaturesCollector {
+    pub async fn collect_signatures(self) -> SignaturesOutcome {
+        _ = self
+            .sign_with_factors() // in decreasing "friction order"
+            .await
+            .inspect_err(|e| eprintln!("Failed to use factor sources: {:#?}", e));
+
+        self.outcome()
+    }
+}
+
+// === PRIVATE ===
+impl SignaturesCollector {
+    /// If all transactions already would fail, or if all transactions already are done, then
+    /// no point in continuing.
+    ///
+    /// `Ok(true)` means "continue", `Ok(false)` means "stop, we are done". `Err(_)` means "stop, we have failed".
+    fn continue_if_necessary(&self) -> Result<bool> {
+        self.state
+            .borrow()
+            .petitions
+            .borrow()
+            .continue_if_necessary()
+    }
+
+    async fn use_factor_sources(&self, factor_sources_of_kind: &FactorSourcesOfKind) -> Result<()> {
+        let interactor = self
+            .dependencies
+            .interactors
+            .interactor_for(factor_sources_of_kind.kind);
+        let factor_sources = factor_sources_of_kind.factor_sources();
+        match interactor {
+            // Parallel Interactor: Many Factor Sources at once
+            SigningInteractor::Parallel(interactor) => {
+                // Prepare the request for the interactor
+                let request = self.request_for_parallel_interactor(
+                    factor_sources
+                        .into_iter()
+                        .map(|f| f.factor_source_id())
+                        .collect(),
+                );
+                let response = interactor.sign(request).await?;
+                self.process_batch_response(response);
+            }
+
+            // Serial Interactor: One Factor Sources at a time
+            // After each factor source we pass the result to the collector
+            // updating its internal state so that we state about being able
+            // to skip the next factor source or not.
+            SigningInteractor::Serial(interactor) => {
+                for factor_source in factor_sources {
+                    // Prepare the request for the interactor
+                    let request =
+                        self.request_for_serial_interactor(&factor_source.factor_source_id());
+
+                    // Produce the results from the interactor
+                    let response = interactor.sign(request).await?;
+
+                    // Report the results back to the collector
+                    self.process_batch_response(response);
+
+                    if !self.continue_if_necessary()? {
+                        break;
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    async fn sign_with_factors_of_kind(
+        &self,
+        factor_sources_of_kind: FactorSourcesOfKind,
+    ) -> Result<()> {
```

Comment:
I think this function does not need to return a Result

Reply:
Thx, FIXED in https://github.com/radixdlt/sargon-mfa/commit/8d9418b3c47d7602455ae664ecf122f0c6a8b6dd